node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,738 @@
+		zap.Uint8("ConsistencyLevel", evt.Msg.ConsistencyLevel),
+	)
+
+	ce.msgC <- evt.Msg
+	messagesConfirmed.WithLabelValues(ce.chainName).Inc()
+}
+
+/*
+This query:
+  `{"all_chain_connections":{}}` is `eyJhbGxfY2hhaW5fY29ubmVjdGlvbnMiOnt9fQ==`
+  which becomes:
+  http://localhost:1319/cosmwasm/wasm/v1/contract/wormhole1nc5tatafv6eyq7llkr2gv50ff9e22mnf70qgjlv737ktmt4eswrq0kdhcj/smart/eyJhbGxfY2hhaW5fY29ubmVjdGlvbnMiOnt9fQ%3D%3D
+
+Returns something like this:
+{
+	"data": {
+		"chain_connections": [
+			[
+				"Y29ubmVjdGlvbi0w",
+				18
+			],
+			[
+				"Y29ubmVjdGlvbi00Mg==",
+				22
+			]
+		]
+	}
+}
+
+*/
+
+type ibcAllChainConnectionsQueryResults struct {
+	Data struct {
+		ChainConnections [][]interface{} `json:"chain_connections"`
+	}
+}
+
+var connectionIdMapQuery = url.QueryEscape(base64.StdEncoding.EncodeToString([]byte(`{"all_chain_connections":{}}`)))
+
+// queryConnectionIdMap queries the contract for the set of chain IDs available on IBC and their corresponding IBC connections.
+func (w *Watcher) queryConnectionIdMap() (map[vaa.ChainID]string, error) {
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	query := fmt.Sprintf(`%s/cosmwasm/wasm/v1/contract/%s/smart/%s`, w.lcdUrl, w.contractAddress, connectionIdMapQuery)
+	connResp, err := client.Get(query)
+	if err != nil {
+		w.logger.Error("channel map query failed", zap.String("query", query), zap.Error(err))
+		return nil, fmt.Errorf("query failed: %w", err)
+	}
+	connBody, err := io.ReadAll(connResp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("read failed: %w", err)
+	}
+	connResp.Body.Close()
+
+	var result ibcAllChainConnectionsQueryResults
+	err = json.Unmarshal(connBody, &result)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response: %s, error: %w", string(connBody), err)
+	}
+
+	if len(result.Data.ChainConnections) == 0 {
+		return nil, fmt.Errorf("query did not return any data")
+	}
+
+	connIdMap := make(map[vaa.ChainID]string)
+	for idx, connData := range result.Data.ChainConnections {
+		if len(connData) != 2 {
+			return nil, fmt.Errorf("connection map entry %d contains %d items when it should contain exactly two, json: %s", idx, len(connData), string(connBody))
+		}
+
+		connectionIdBytes, err := base64.StdEncoding.DecodeString(connData[0].(string))
+		if err != nil {
+			return nil, fmt.Errorf("connection ID for entry %d is invalid base64: %s, err: %s", idx, connData[0], err)
+		}
+
+		connectionID := string(connectionIdBytes)
+		chainID := vaa.ChainID(connData[1].(float64))
+		connIdMap[chainID] = connectionID
+		w.logger.Debug("IBC connection ID mapping", zap.String("connectionID", connectionID), zap.Uint16("chainID", uint16(chainID)))
+	}
+
+	return connIdMap, nil
+}
+
+// getConnectionID returns the IBC connection ID associated with the given IBC channel. It uses a cache to avoid constantly
+// querying worm chain. This works because once an IBC channel is closed its ID will never be reused.
```

Comment:
Might be good to link to the relevant IBC specs here for future reference.
- Connection spec stating that connections can never be closed and identifiers never re-allocated: https://github.com/cosmos/ibc/tree/main/spec/core/ics-003-connection-semantics#sub-protocols
- Channel spec stating channels cannot be reopened and identifiers not re-used: https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics#closing-handshake